src/engine/gnc-pricedb.c

```diff
@@ -1345,68 +1356,299 @@ pricedb_remove_foreach_pricelist (gpointer key,
-        node = g_list_next(node);
-
     /* now check each item in the list */
     g_list_foreach(node, (GFunc)check_one_price_date, data);
 
     LEAVE(" ");
 }
 
+static gint
+compare_prices_by_commodity_date (gconstpointer a, gconstpointer b)
+{
+    Timespec time_a;
+    Timespec time_b;
+    gnc_commodity *comma;
+    gnc_commodity *commb;
+    gnc_commodity *curra;
+    gnc_commodity *currb;
+    gint result;
+
+    if (!a && !b) return 0;
+    /* nothing is always less than something */
+    if (!a) return -1;
+
+    comma = gnc_price_get_commodity ((GNCPrice *) a);
+    commb = gnc_price_get_commodity ((GNCPrice *) b);
+
+    if (!gnc_commodity_equal(comma, commb))
+        return gnc_commodity_compare(comma, commb);
+
+    curra = gnc_price_get_currency ((GNCPrice *) a);
+    currb = gnc_price_get_currency ((GNCPrice *) b);
+
+    if (!gnc_commodity_equal(curra, currb))
+        return gnc_commodity_compare(curra, currb);
+
+    time_a = gnc_price_get_time((GNCPrice *) a);
+    time_b = gnc_price_get_time((GNCPrice *) b);
+
+    result = -timespec_cmp(&time_a, &time_b);
+    if (result) return result;
+
+    /* For a stable sort */
+    return guid_compare (gnc_price_get_guid((GNCPrice *) a),
+                         gnc_price_get_guid((GNCPrice *) b));
+}
+
+static gboolean
+price_commodity_and_currency_equal (GNCPrice *a, GNCPrice *b)
+{
+    gboolean ret_comm = FALSE;
+    gboolean ret_curr = FALSE;
+
+    if (gnc_commodity_equal (gnc_price_get_commodity(a), gnc_price_get_commodity (b)))
+        ret_comm = TRUE;
+
+    if (gnc_commodity_equal (gnc_price_get_currency(a), gnc_price_get_currency (b)))
+        ret_curr = TRUE;
+
+    return (ret_comm && ret_curr);
+}
+
 static void
-pricedb_remove_foreach_currencies_hash (gpointer key,
-                                        gpointer val,
-                                        gpointer user_data)
+gnc_pricedb_remove_old_prices_pinfo (GNCPrice *price, gboolean keep_message)
 {
-    GHashTable *currencies_hash = (GHashTable *) val;
+    GDate price_date = timespec_to_gdate (gnc_price_get_time (price));
+    char date_buf[MAX_DATE_LENGTH+1];
 
-    ENTER("key %p, value %p, data %p", key, val, user_data);
-    g_hash_table_foreach(currencies_hash,
-                         pricedb_remove_foreach_pricelist, user_data);
-    LEAVE(" ");
+    if (g_date_valid (&price_date))
+    {
+        qof_print_gdate (date_buf, MAX_DATE_LENGTH, &price_date);
+
+        if (keep_message)
+        {
+            PINFO("#### Keep price with date %s, commodity is %s, currency is %s", date_buf,
+                     gnc_commodity_get_printname(gnc_price_get_commodity(price)),
+                     gnc_commodity_get_printname(gnc_price_get_currency(price)));
+        }
+        else
+            PINFO("## Remove price with date %s", date_buf);
+    }
+    else
+        PINFO("Keep price date is invalid");
+}
+
+static GNCPrice*
+save_cloned_price (GNCPrice *price, GNCPrice *clone_price)
```

Comment:
I considered some thing like this when first implementing but went the wrong way.